Client plugin system & Server plugin rework

dzervas opened the pull request:
Adds some basic support for client plugins. Currently they should do whatever they do in `__init__`.
Refined the server plugin system, so that the commands' docstring is parsed for help & syntax. Also the functions that are commands just need the `@command(states)` decorator. I've patched the plugins to work with the new system.

dzervas committed:
Removed gitignore

Commit:
Merge branch 'master' of github.com:panagiks/rspet

Commit:
Add basic plugin support on client

Commit:
Add plugin unloading on client

Commit:
Refactored most of the server plugin system.

Commit:
Test if command comes from specific plugin before loading it

Commit:
Add docstring parsing

Commit:
Fix broken help in API

Commit:
Update client plugin system to match server updates

Commit:
Add command decorator to client plugin system

Commit:
Make autocompletion complete all plugins' commands and make them
compatible with the previous, camel case commands

panagiks commented:
Since it's a relatively big PR I'll have to test it a bit more thoroughly, so I apologize in advance for the delay. If anyone else wants/has time to test feel free to post your comments :) Thanks for the additions @dzervas !

dzervas commented:
Take your time :) (I'll try to keep them smaller, as they should be to make things easier on both sides)
What we need is some unit tests. I kinda hate writing them, but I'll see what I can do.

panagiks commented:
Everything seems to be working perfectly :) And the use of decorators for the plugin functions will make them a lot easier to develop and keep track of. Auto-completion seems to be case sensitive (although ignore-case is set to on) and uses the camel-cased version of the commands whereas the help displays an all lower case version of the commands (as gathered from the function names). This does not cause any functional problems as the command interpreter will accept both camel-cased and lower-case commands, it only causes a minor confusion (potentially) but it's nothing to worry about I'll get it sorted in a couple of days.

Only one thing before I merge it. Could you remove the prints from the client module so that it's "production ready".

Thank you for yet another great addition :)

panagiks committed:
Removed prints from client, added reporting

Client now reports successful and failed plugin load attempts to the server, which in turn keeps a "memory" of the loaded plugins on every client.
